### ISSUE TITLE
Exception in checkIndexConstraintForProperty when handling assignment to property in constructor

### ISSUE THREAD

**Issue by TysonAndre** [Microsoft/TypeScript#15681]
**TypeScript Version:**  **master branch as of today(may 9th)** (https://github.com/Microsoft/TypeScript/commit/a72a9aa5d5777bd99e00411ccf5b920e5dc675d0)

**Code**

```ts
// (Using react version "15.4.2")
// This is a .js file, using --checkJS (and extra lines were cut out)
// A *self-contained* demonstration of the problem follows...
// 
define(function(require) {
    'use strict';

    const React = require('react');
    const { Component } = React;

    class Something extends Component {

        constructor(props) {
            super(props)

            this.state = { };  // This overrides a parent property - see https://facebook.github.io/react/docs/react-component.html#state
        }
    }

    return Something;
});
```

Command used (NOTE: there was no config file used here, just command line flags)

```bash
tsc --checkJs --noEmit --allowJs example.js
```

**Expected behavior:**
Does not crash
**Actual behavior:**
Crashes because of the assignment of `{}` to this.state (property of parent class Component)

```
TypeError: Cannot read property 'kind' of undefined
    at checkIndexConstraintForProperty (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:36208:61)
    at /usr/local/node/lib/node_modules/typescript/lib/tsc.js:36173:21
    at Object.forEach (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:288:30)
    at checkIndexConstraints (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:36171:20)
    at checkClassLikeDeclaration (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:36406:17)
    at checkClassDeclaration (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:36324:13)
    at checkSourceElement (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:37259:28)
    at Object.forEach (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:288:30)
    at checkBlock (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:35383:16)
    at checkSourceElement (/usr/local/node/lib/node_modules/typescript/lib/tsc.js:37222:28)
```

This was seen after the fix for https://github.com/Microsoft/TypeScript/issues/15616 was merged
This happens whether or not PR 15594 is merged (I think)

**Comment by RyanCavanaugh** [comment]
Looks extremely similar to #15525

**Comment by mhegazy** [comment]
indeed a duplicate. the fix should in tomorrows' `typescript@next`.

**Comment by TysonAndre** [comment]
It isn't a duplicate, this happened after the fix was merged.

See https://github.com/Microsoft/TypeScript/pull/15586/files#r115359347 for the line causing this error

**Comment by sandersn** [comment]
I believe this class of bugs should be squashed completely with #15594, which prevents access to `Declaration.name` and adds a `getNameOfDeclaration` utility. It also introduces a new type, `NamedDeclaration`, for the declarations that always have a `name`.

**Comment by sandersn** [comment]
Fix is ready #15594

**Comment by ninja-** [comment]
is there any workaround for now?

**Comment by sandersn** [comment]
The workaround is not to use checkjs. I just merged the fix, so it should be in tomorrow's nightly (May 11)

**Comment by chrisleck** [comment]
Any word on this will be released (presumably the release date of 2.3.3)?

**Comment by mhegazy** [comment]
> Any word on this will be released (presumably the release date of 2.3.3)?

later this week. `typescript@2.3.3-insiders.20170512` should be what we are planning to release in 2.3.3
